File: app/src/main/java/com/arnaudpiroelle/manga/core/provider/ProviderRegistry.java
package com.arnaudpiroelle.manga.core.provider;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ProviderRegistry {

    Map<String, MangaProvider> registry;

    public ProviderRegistry() {
        registry = new HashMap<>();
    }

    private void register(MangaProvider mangaProvider) {
        registry.put(mangaProvider.getName(), mangaProvider);
    }

    public MangaProvider get(String providerName) {
        return registry.get(providerName);
    }

    public List<MangaProvider> list(){
        return new ArrayList<>(registry.values());
    }

    public static class ProviderRegistryBuilder {

        private ProviderRegistry providerRegistry;

        public static ProviderRegistryBuilder createProviderRegister() {
            return new ProviderRegistryBuilder();
        }

        public ProviderRegistryBuilder() {
            providerRegistry = new ProviderRegistry();
        }

        public ProviderRegistryBuilder withProvider(MangaProvider... mangaProviders) {
            for (MangaProvider mangaProvider : mangaProviders) {
                providerRegistry.register(mangaProvider);
            }

            return this;
        }

        public ProviderRegistry build() {
            return providerRegistry;
        }

    }
}

File: app/src/main/kotlin/com/arnaudpiroelle/manga/core/provider/ProviderRegistry.kt
package com.arnaudpiroelle.manga.core.provider

import java.util.*

open class ProviderRegistry {

    private var registry: MutableMap<String, MangaProvider> = HashMap()

    fun register(name: String, mangaProvider: MangaProvider) {
        registry[name] = mangaProvider
    }

    fun find(providerName: String): MangaProvider {
        return registry[providerName]!!
    }

    fun list(): Map<String, MangaProvider> {
        return registry
    }


}

class ProviderRegistryBuilder {

    private val providerRegistry: ProviderRegistry = ProviderRegistry()

    fun withProvider(name: String, mangaProvider: MangaProvider): ProviderRegistryBuilder {
        providerRegistry.register(name, mangaProvider)

        return this
    }

    fun build(): ProviderRegistry {
        return providerRegistry
    }

    companion object {

        fun createProviderRegister(): ProviderRegistryBuilder {
            return ProviderRegistryBuilder()
        }
    }

}